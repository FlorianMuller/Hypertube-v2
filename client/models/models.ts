import { AxiosResponse, AxiosError, Canceler } from "axios";

export interface Locale {
  locale: string;
  setLocale: (locale: string) => void;
}

export type Method = "get" | "put" | "post" | "patch" | "delete";
export type Headers = { [key: string]: string } | undefined;
export type Data = object | undefined;

export interface UseApiOption {
  method?: Method;
  headers?: Headers;
  data?: Data;
  hotReload?: boolean;
  validateStatus?: (status: number) => boolean;
}

export interface UseApiReturn<T, E> {
  callApi: (freshData?: Data) => Promise<void>;
  loading: boolean;
  res: AxiosResponse<T>;
  resData: T;
  error: AxiosError<E>;
  setUrl: (url: string) => void;
  setMethod: (method: Method) => void;
  setHeaders: (headers: Headers) => void;
  setData: (data: Data) => void;
  cancelAllRequests: Canceler;
}

export interface ApiAuthResponse {
  validToken: boolean;
  error?: string;
}

export interface AuthInfo {
  username: string;
  password: string;
  [key: string]: string;
}

export interface Comment {
  movieName: string;
  date: number;
  name: string;
  body: string;
  stars: number;
  _id: number;
}

export interface Review {
  id: string;
  name: string;
  date: number;
  stars: number;
  body: string;
}

export interface Filters {
  query: string;
  collections: string[];
  year: number;
  minRating: number;
}

export interface Movie {
  id: string;
  title: string;
  cover: string;
  year: number;
  summary: string;
  genres: string[];
  rating: number;
  runtime: number;
}

export interface ApiSearchReponse {
  movies: Movie[];
  nextPage: boolean;
}

export type CustomSnackbarVariant =
  | "success"
  | "warning"
  | "error"
  | "info"
  | "default";
