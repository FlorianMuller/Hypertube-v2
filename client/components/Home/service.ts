export const getPictureSize = (
  isXs: boolean
): { width: number; height: number } => {
  return isXs ? { width: 300, height: 450 } : { width: 375, height: 562.5 };
};

export default { getPictureSize };
