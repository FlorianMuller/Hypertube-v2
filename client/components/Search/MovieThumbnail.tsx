import React, { ReactElement } from "react";
import { Link } from "react-router-dom";
import { useIntl } from "react-intl";
import { Button, Chip, Typography, Card } from "@material-ui/core";
import Image from "material-ui-image";
import StarIcon from "@material-ui/icons/Star";

import useSearchStyles from "./Search.styles";
import { Movie } from "../../models/models";

interface Props {
  movie: Movie;
}

const MovieThumbnail = ({ movie }: Props): ReactElement => {
  const classes = useSearchStyles({});
  const { formatMessage: _t } = useIntl();

  return (
    <div className={classes.thumbnailContainer} key={movie.id}>
      <Image
        animationDuration={500}
        src={movie.cover || "no-image"}
        color="rgba(0,0,0,0)"
        imageStyle={{ width: 300, height: 450 }}
        style={{ width: 300, height: 450 }}
        disableSpinner
        errorIcon={
          <Card className={classes.altContainer}>
            <Typography variant="h4">{movie.title}</Typography>
          </Card>
        }
      />
      <div className={classes.thumbnailOverlay}>
        <Typography variant="h5">{movie.title}</Typography>
        <Typography variant="h6">{movie.year}</Typography>
        <Typography variant="caption" className={classes.summary}>
          {movie.summary}
        </Typography>
        <div className={classes.metaInfos}>
          {movie.genres && (
            <div>
              {movie.genres.map((genre: string) => (
                <Chip
                  className={classes.tag}
                  label={genre}
                  clickable
                  color="primary"
                  key={genre}
                />
              ))}
            </div>
          )}
          <Typography className={classes.rating}>
            {movie.rating !== null && (
              <>
                {movie.rating} <StarIcon className={classes.ratingIcon} />
              </>
            )}
            {movie.rating !== null && movie.runtime && <span> - </span>}
            {movie.runtime && <span>{movie.runtime} mins</span>}
          </Typography>
          <Link to={`/movie/${movie.id}`} className={classes.watchLink}>
            <Button color="primary" variant="contained">
              {_t({ id: "search.film.watch" })}
            </Button>
          </Link>
        </div>
      </div>
    </div>
  );
};

export default MovieThumbnail;
