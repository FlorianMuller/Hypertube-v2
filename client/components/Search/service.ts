import qs from "qs";
import moment from "moment";

export const clamp = (str: string, maxLength: number): string =>
  str
    ? `${str.slice(0, maxLength)}${str.length <= maxLength ? "" : "..."}`
    : "";

export const formatQueryUrl = (search: string, page: number): string => {
  const {
    query,
    collections,
    startYear,
    endYear,
    minRating,
    maxRating
  } = qs.parse(search.slice(1));

  return `/search?${qs.stringify({
    query: query || undefined,
    collections: collections || undefined,
    startYear:
      startYear === 1900 && endYear === moment().year() ? undefined : startYear,
    endYear:
      startYear === 1900 && endYear === moment().year() ? undefined : endYear,
    minRating: minRating === 0 && maxRating === 5 ? undefined : minRating,
    maxRating: minRating === 0 && maxRating === 5 ? undefined : maxRating,
    page
  })}`;
};
