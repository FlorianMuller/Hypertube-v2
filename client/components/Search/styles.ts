import { makeStyles } from "@material-ui/core/styles";

export const useSearchStyles = makeStyles((theme) => ({
  snackbar: {
    bottom: theme.spacing(5),
    right: theme.spacing(5),
    position: "fixed",
    width: "200px"
  },
  snackbarLoading: {
    bottom: theme.spacing(5),
    right: theme.spacing(0),
    position: "fixed",
    width: "200px"
  },
  thumbsContainer: {
    border: "3px solid red",
    overflow: "auto",
    position: "relative",
    display: "flex",
    alignItems: "center",
    padding: theme.spacing(4),
    paddingBottom: 0,
    zIndex: 0
  }
}));

export const useFilmStyles = makeStyles((theme) => ({
  container: {
    height: "100%",
    width: "100%",
    display: "flex",
    flexDirection: "column",
    textAlign: "center",
    padding: theme.spacing(5)
  },
  thumbnail: {
    marginTop: theme.spacing(3),
    alignSelf: "center"
  },
  ratingContainer: {
    display: "flex",
    justifyContent: "center",
    alignItems: "self-end",
    margin: `${theme.spacing(1)} 0`
  },
  descriptionContainer: {
    textAlign: "justify",
    marginTop: theme.spacing(3),
    marginBottom: theme.spacing(2),
    paddingRight: theme.spacing(1),
    height: "30vh",
    overflowY: "auto"
  },
  tag: {
    margin: theme.spacing(0.5)
  },
  buttonContainer: {
    marginTop: theme.spacing(4),
    alignSelf: "center",
    width: "30%"
  }
}));

export const useThumbnailStyles = makeStyles((theme) => ({
  container: {
    cursor: "pointer",
    display: "flex",
    padding: `${theme.spacing(2)}px 0`,
    transformOrigin: "0% 50%",
    transition: "transform 0.2s",
    "&:hover": {
      transform: "scale(1.15)",
      backgroundColor: "rgba(255, 255, 255, 0.1)"
    }
  },
  content: {
    display: "flex",
    flexDirection: "column",
    paddingLeft: theme.spacing(1)
  },
  filmInfo: {
    lineHeight: "initial",
    display: "flex",
    alignItems: "flex-end"
  }
}));
