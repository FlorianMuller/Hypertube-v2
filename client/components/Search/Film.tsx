import React, { ReactElement } from "react";
import moment from "moment";
import _ from "lodash";
import { useHistory } from "react-router-dom";
import { useIntl } from "react-intl";

import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import Rating from "@material-ui/lab/Rating";
import Chip from "@material-ui/core/Chip";

import { Film as FilmInterface } from "../../models/models";

import { useFilmStyles } from "./styles";

interface Props {
  film: FilmInterface;
}

const Film = ({ film }: Props): ReactElement => {
  const classes = useFilmStyles({});
  const history = useHistory();
  const { formatMessage: _t } = useIntl();

  const searchTag = (subj: string): void =>
    history.push({
      state: {
        ...history.location.state,
        query: subj,
        previousHistory: true
      }
    });

  const renderTags = (): ReactElement[] =>
    _.flatten(Array(film.subject)).map((subj: string) => (
      <Chip
        key={subj}
        className={classes.tag}
        label={subj}
        clickable
        color="primary"
        onClick={(): void => searchTag(subj)}
      />
    ));

  return (
    <div className={classes.container}>
      <Typography variant="h3">{film.title}</Typography>
      <img
        className={classes.thumbnail}
        alt="thumb"
        src={`https://archive.org/download/${film.identifier}/__ia_thumb.jpg`}
      />
      <div className={classes.ratingContainer}>
        <Rating
          name="read-only"
          value={Number(film.avg_rating)}
          precision={0.01}
          readOnly
        />
        <div>
          {film.avg_rating || `- ${_t({ id: "search.film.norating" })}`}
        </div>
      </div>
      <span>
        {`${_t(
          { id: "search.film.author" },
          { author: film.creator || _t({ id: "search.film.unknown_author" }) }
        )}`}{" "}
        - {moment(film.date).year()}
      </span>
      <Typography variant="caption" className={classes.descriptionContainer}>
        {film.description}
      </Typography>
      <div>{_t({ id: "search.film.tags" })} : </div>
      <div>{renderTags()}</div>
      <div className={classes.buttonContainer}>
        <Button
          color="primary"
          variant="contained"
          fullWidth
          onClick={(): void => history.push(`/movies/${film.identifier}`)}
        >
          {_t({ id: "search.film.watch" })}
        </Button>
      </div>
    </div>
  );
};

export default Film;
