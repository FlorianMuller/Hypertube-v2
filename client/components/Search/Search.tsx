import React, { ReactElement, useState, useEffect } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { useIntl } from "react-intl";
import _ from "lodash";
import InfiniteScroll from "react-infinite-scroller";

import Skeleton from "@material-ui/lab/Skeleton";
import CircularProgress from "@material-ui/core/CircularProgress";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";

import useApi from "../../hooks/useApi";
import { formatQueryUrl } from "./Search.service";
import { Movie, ApiSearchReponse } from "../../models/models";
import useSearchStyles from "./Search.styles";

import MovieThumbnail from "./MovieThumbnail";

const Search = (): ReactElement => {
  const classes = useSearchStyles({});
  const { formatMessage: _t } = useIntl();
  const [printedMovieList, setPrintedMovieList] = useState<Movie[]>([]);
  const [movieList, setMovieList] = useState<Movie[]>([]);
  const [page, setPage] = useState(1);
  const history = useHistory();
  const location = useLocation();

  const { resData: data, loading, error, setUrl } = useApi<
    ApiSearchReponse,
    void
  >(formatQueryUrl(location.search, 1), {
    hotReload: true
  });

  /**
   * Starting a new search
   */
  useEffect(() => {
    setPage(1);
    setMovieList([]);
    setPrintedMovieList([]);
    setUrl(formatQueryUrl(location.search, 1));
  }, [location.search, location.pathname]);

  /**
   * New Data receive
   */
  useEffect(() => {
    if (data && data.movies?.length) {
      if (page === 1) {
        setPrintedMovieList(data.movies.slice(0, 12));
        setMovieList(data.movies.slice(12));
      } else {
        setMovieList((oldMovieList) => [...oldMovieList, ...data.movies]);
      }
    }
  }, [data]);

  /**
   * Getting next 12 movies (and next page if needed)
   */
  const loadMore = (): void => {
    // Getting next page if we don't have enough moovie in stock
    if (!loading && movieList.length < 24) {
      setPage(page + 1);
      setUrl(formatQueryUrl(location.search, page + 1));
    }

    // Printing next 12 movies
    if (movieList.length) {
      setPrintedMovieList((old) => [...old, ...movieList.slice(0, 12)]);
      setMovieList((old) => old.slice(12));
    }
  };

  if (error) {
    history.push("/error");
  }

  return (
    <Container maxWidth="xl">
      <InfiniteScroll
        className={classes.container}
        initialLoad={false}
        loadMore={loadMore}
        hasMore={!!movieList.length}
        threshold={100}
      >
        {/* No media found */}
        {!printedMovieList.length && !loading && (
          <div className={classes.noMediaContainer}>
            <img
              src={`${window.location.origin}/public/no-media.png`}
              alt="no-media"
              className={classes.noMedia}
            />
            <Typography variant="h5">
              {_t({ id: "search.no_media" })}
            </Typography>
          </div>
        )}

        {/* Medias list */}
        {printedMovieList.map((movie) => (
          <MovieThumbnail key={movie.id} movie={movie} />
        ))}

        {/* Fake movies Loading */}
        {loading &&
          movieList.length === 0 &&
          _.times(12, (i) => (
            <div className={classes.thumbnailContainer} key={i}>
              <Skeleton
                variant="rect"
                width={300}
                height={450}
                className={classes.skeletonThumbnail}
              />
              <div className={classes.skeletonProgress}>
                <CircularProgress />
              </div>
            </div>
          ))}
      </InfiniteScroll>
    </Container>
  );
};

export default Search;
