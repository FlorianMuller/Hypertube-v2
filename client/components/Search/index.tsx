import React, { ReactElement, useState, useEffect } from "react";
import qs from "qs";
import { useHistory } from "react-router-dom";
import { useIntl } from "react-intl";

import CircularProgress from "@material-ui/core/CircularProgress";
import Typography from "@material-ui/core/Typography";
import Paper from "@material-ui/core/Paper";
import InfiniteScroll from "react-infinite-scroller";

import Thumbnail from "./Thumbnail";
import Film from "./Film";

import useApi from "../../hooks/useApi";

import { formatUrl } from "./service";

import { useSearchStyles } from "./styles";

const Search = (): ReactElement => {
  const classes = useSearchStyles({});
  const history = useHistory();
  const { formatMessage: _t } = useIntl();
  const [filmData, setFilmData] = useState(null);
  const [filmList, setFilmList] = useState([]);
  const [block, setBlock] = useState(false);
  const [page, setPage] = useState(1);
  const { data, loading, error, setUrl } = useApi(
    `http://localhost:8080/API/getFilms?${qs.stringify({
      ...history.location.state,
      page: 1
    })}`
  );

  useEffect(() => {
    if (history.location.state.previousHistory && !loading && !block) {
      setPage(1);
      setFilmList([]);
      setUrl(formatUrl(history.location.state, 1));
    }
  }, [history.location.state]);

  useEffect(() => {
    if (data?.docs) {
      const docs = data.docs as Record<string, unknown>[];

      setFilmList(() => [...filmList, ...docs]);
      setBlock(false);
    }
  }, [data]);

  const loadMore = (): void => {
    setBlock(true);
    setPage(page + 1);
    setUrl(formatUrl(history.location.state, page + 1));
  };

  if (error) {
    history.push("/error");
  }

  return (
    <div className={classes.container}>
      <div className={classes.resultContainer}>
        <div className={classes.thumbsContainer}>
          <Paper className={classes.infoSidebar}>
            {loading ? (
              <>
                {_t({ id: "search.loading" })}
                <CircularProgress />
              </>
            ) : (
              <Typography>
                {_t(
                  { id: "search.film_count" },
                  { count: Number(data.numFound || 0) }
                )}
              </Typography>
            )}
          </Paper>
          <InfiniteScroll
            initialLoad={false}
            useWindow={false}
            loadMore={loadMore}
            hasMore={!block && !loading && page * 10 < data.numFound}
          >
            {filmList.map((film) => (
              <Thumbnail
                key={film.identifier}
                film={film}
                onClick={(): void => setFilmData(film)}
              />
            ))}
          </InfiniteScroll>
        </div>
        <div className={classes.filmContainer}>
          {filmData ? (
            <Film film={filmData} />
          ) : (
            <Typography variant="h2">
              {_t({ id: "search.details_placeholder" })}
            </Typography>
          )}
        </div>
      </div>
    </div>
  );
};

export default Search;
