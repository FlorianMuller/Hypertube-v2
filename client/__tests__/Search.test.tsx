import React from "react";
import Adapter from "enzyme-adapter-react-16";
import EnzymeToJson from "enzyme-to-json";
import { configure } from "enzyme";
import { History, Location, LocationState } from "history";

import Search from "../components/Search";

import { mountWithIntl } from "./helpers/intl-enzyme-test-helper";

import * as ServiceSearch from "../components/Search/service";
import { UseApiReturn, ApiSearchReponse } from "../models/models";

configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: (): History<LocationState> => ({
    length: null,
    action: null,
    location: { pathname: "/search", search: "", state: {}, hash: null },
    push: jest.fn(),
    replace: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    block: jest.fn(),
    listen: jest.fn(),
    createHref: jest.fn()
  }),
  useLocation: (): Location<LocationState> => ({
    pathname: "/search",
    search: "",
    state: {},
    hash: null
  })
}));

jest.mock("../hooks/useApi", () => (): UseApiReturn<
  ApiSearchReponse,
  void
> => ({
  callApi: jest.fn(),
  loading: false,
  res: {
    data: { movies: [], nextPage: false },
    status: 200,
    statusText: null,
    headers: null,
    config: null
  },
  resData: { movies: [], nextPage: false },
  error: null,
  setUrl: jest.fn(),
  setMethod: jest.fn(),
  setHeaders: jest.fn(),
  setData: jest.fn(),
  cancelAllRequests: jest.fn()
}));

describe("Search", () => {
  describe("search en", () => {
    it("renders correctly", () => {
      const domNode = mountWithIntl(<Search />, "en");
      expect(EnzymeToJson(domNode)).toMatchSnapshot();
    });
  });

  describe("search fr", () => {
    it("renders correctly", () => {
      const domNode = mountWithIntl(<Search />, "fr");
      expect(EnzymeToJson(domNode)).toMatchSnapshot();
    });
  });

  describe("Service", () => {
    describe("formatQueryUrl", () => {
      it("should get a full query", () => {
        const testQuery = ServiceSearch.formatQueryUrl(
          "?query=top&collections%5B0%5D=collec1&collections%5B1%5D=collec2",
          1
        );

        expect(testQuery).toBe(
          "/movies?query=top&collections%5B0%5D=collec1&collections%5B1%5D=collec2&page=1"
        );
      });

      it("should get a empty query", () => {
        const testQuery = ServiceSearch.formatQueryUrl("", 1);
        expect(testQuery).toBe("/movies?page=1");
      });
    });
  });
});
