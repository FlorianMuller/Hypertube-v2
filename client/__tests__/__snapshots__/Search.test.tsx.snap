// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search search en renders correctly 1`] = `
<Search>
  <WithStyles(ForwardRef(Container))
    maxWidth="xl"
  >
    <ForwardRef(Container)
      classes={
        Object {
          "disableGutters": "MuiContainer-disableGutters",
          "fixed": "MuiContainer-fixed",
          "maxWidthLg": "MuiContainer-maxWidthLg",
          "maxWidthMd": "MuiContainer-maxWidthMd",
          "maxWidthSm": "MuiContainer-maxWidthSm",
          "maxWidthXl": "MuiContainer-maxWidthXl",
          "maxWidthXs": "MuiContainer-maxWidthXs",
          "root": "MuiContainer-root",
        }
      }
      maxWidth="xl"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthXl"
      >
        <InfiniteScroll
          className="makeStyles-container-1"
          element="div"
          getScrollParent={null}
          hasMore={false}
          initialLoad={false}
          isReverse={false}
          loadMore={[Function]}
          loader={null}
          pageStart={0}
          ref={null}
          threshold={250}
          useCapture={false}
          useWindow={true}
        >
          <div
            className="makeStyles-container-1"
          />
        </InfiniteScroll>
      </div>
    </ForwardRef(Container)>
  </WithStyles(ForwardRef(Container))>
</Search>
`;

exports[`Search search fr renders correctly 1`] = `
<Search>
  <WithStyles(ForwardRef(Container))
    maxWidth="xl"
  >
    <ForwardRef(Container)
      classes={
        Object {
          "disableGutters": "MuiContainer-disableGutters",
          "fixed": "MuiContainer-fixed",
          "maxWidthLg": "MuiContainer-maxWidthLg",
          "maxWidthMd": "MuiContainer-maxWidthMd",
          "maxWidthSm": "MuiContainer-maxWidthSm",
          "maxWidthXl": "MuiContainer-maxWidthXl",
          "maxWidthXs": "MuiContainer-maxWidthXs",
          "root": "MuiContainer-root",
        }
      }
      maxWidth="xl"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthXl"
      >
        <InfiniteScroll
          className="makeStyles-container-1"
          element="div"
          getScrollParent={null}
          hasMore={false}
          initialLoad={false}
          isReverse={false}
          loadMore={[Function]}
          loader={null}
          pageStart={0}
          ref={null}
          threshold={250}
          useCapture={false}
          useWindow={true}
        >
          <div
            className="makeStyles-container-1"
          />
        </InfiniteScroll>
      </div>
    </ForwardRef(Container)>
  </WithStyles(ForwardRef(Container))>
</Search>
`;
